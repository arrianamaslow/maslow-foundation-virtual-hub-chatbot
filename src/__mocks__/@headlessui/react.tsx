/**
 * @jest-environment jsdom
 */

import { DialogProps } from '@headlessui/react'
import React, { ReactNode } from 'react'

type DialogPanelTestProps = {
  children?: ReactNode
}

interface DialogTestProps extends DialogProps {
  children?: ReactNode
}

const DialogPanel: React.FC<DialogPanelTestProps> = ({ children }) => {
  return <div data-testid='mock-dialog-panel-component'>{children}</div>
}

const Dialog: React.FC<DialogTestProps> = ({ open, children }) => {
  return (
    <div data-testid='mock-dialog-component'>
      {open ? 'DIALOGUEOPENED' : 'DIALOGUECLOSED'}
      {children}
    </div>
  )
}

export { Dialog, DialogPanel }
