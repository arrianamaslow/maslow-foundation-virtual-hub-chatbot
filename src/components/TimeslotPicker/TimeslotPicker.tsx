// src/components/TimeSelector.tsx
import React from 'react'
import Button from '../Button/Button'
type TimeslotPickerProps = {
  date: Date
  times: string[]
  selectedTime: string | null
  handleSelection: (time: string) => void
}

const TimeslotPicker: React.FC<TimeslotPickerProps> = ({
  times,
  date,
  selectedTime,
  handleSelection
}) => {
  const numOfRows = Math.ceil(times.length / 2)
  const optionsWeekday: Intl.DateTimeFormatOptions = { weekday: 'long' }
  const optionsDay: Intl.DateTimeFormatOptions = { day: 'numeric' }
  const optionsMonth: Intl.DateTimeFormatOptions = { month: 'long' }
  const optionsYear: Intl.DateTimeFormatOptions = { year: 'numeric' }

  const weekday = new Intl.DateTimeFormat('en-GB', optionsWeekday).format(date)
  const day = new Intl.DateTimeFormat('en-GB', optionsDay).format(date)
  const month = new Intl.DateTimeFormat('en-GB', optionsMonth).format(date).toUpperCase()
  const year = new Intl.DateTimeFormat('en-GB', optionsYear).format(date)

  const dayMonth = `${day} ${month}`

  return (
    <div className='flex w-full flex-col justify-start gap-y-[48px] md:gap-y-[80px]'>
      <p className='font-baskervald text-[1.625rem] md:text-[3.125rem]'>{`${weekday}, ${dayMonth} ${year}`}</p>
      <div className='flex w-full'>
        <div
          className={`ml-auto gap-x-8 gap-y-[16px] grid grid-flow-col md:grid-rows-none md:ml-0 md:grid-flow-row md:grid-cols-5 md:w-full ${getGridRowStyle(numOfRows)}`}
        >
          {times.map((time, i) => (
            <div className='w-fit' key={i}>
              <Button
                text={time}
                isDisabled={false}
                buttonType={selectedTime === time ? `primary` : `secondary`}
                onClick={() => {
                  handleSelection(time)
                }}
              />
            </div>
          ))}
        </div>
      </div>
    </div>
  )
}

function getGridRowStyle(num: number): string {
  const rowStyle = [
    'grid-rows-1',
    'grid-rows-2',
    'grid-rows-3',
    'grid-rows-4',
    'grid-rows-5',
    'grid-rows-6',
    'grid-rows-7',
    'grid-rows-8',
    'grid-rows-9',
    'grid-rows-10'
  ]
  if (num > 10) {
    return rowStyle[rowStyle.length - 1]
  }
  return rowStyle[num - 1]
}

export default TimeslotPicker
