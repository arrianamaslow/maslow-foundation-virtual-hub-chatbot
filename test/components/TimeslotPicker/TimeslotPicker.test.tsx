/**
 * @jest-environment jsdom
 */

import TimeslotPicker from '@/components/TimeslotPicker/TimeslotPicker'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

jest.mock('@/components/Button/Button', () => {
  const MockButton = ({ text, onClick, buttonType }: any) => (
    <button onClick={onClick} data-testid={text} className={buttonType}></button>
  )
  return MockButton
})

const timesExample = [
  '08:00',
  '09:00',
  '10:00',
  '11:00',
  '12:00',
  '13:00',
  '14:00',
  '15:00',
  '16:00',
  '17:00'
]

describe('<TimeslotPicker />', () => {
  it('renders correctly', () => {
    const tree = render(
      <TimeslotPicker
        times={timesExample}
        date={new Date(2024, 6, 6)}
        selectedTime={''}
        handleSelection={() => {}}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('has a button for every time in the timesExample list', () => {
    render(
      <TimeslotPicker
        times={timesExample}
        date={new Date(Date.now())}
        selectedTime={''}
        handleSelection={() => {}}
      />
    )
    timesExample.forEach((time) => {
      const button = screen.getByTestId(time)
      expect(button).toBeInTheDocument()
    })
  })

  it('has buttons which are clickable', () => {
    const timeSetter = jest.fn()
    render(
      <TimeslotPicker
        times={timesExample}
        date={new Date(Date.now())}
        selectedTime={''}
        handleSelection={timeSetter}
      />
    )
    timesExample.forEach((time) => {
      const button = screen.getByTestId(time)
      fireEvent.click(button)
      expect(timeSetter).toHaveBeenCalled()
    })
  })
})
