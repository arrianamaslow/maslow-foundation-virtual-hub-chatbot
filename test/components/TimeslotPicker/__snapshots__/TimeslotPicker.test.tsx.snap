// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeslotPicker /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full flex-col justify-start gap-y-[48px] md:gap-y-[80px]"
      >
        <p
          class="font-baskervald text-[1.625rem] md:text-[3.125rem]"
        >
          Saturday, 6 JULY 2024
        </p>
        <div
          class="flex w-full"
        >
          <div
            class="ml-auto gap-x-8 gap-y-[16px] grid grid-flow-col md:grid-rows-none md:ml-0 md:grid-flow-row md:grid-cols-5 md:w-full grid-rows-5"
          >
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="08:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="09:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="10:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="11:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="12:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="13:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="14:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="15:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="16:00"
              />
            </div>
            <div
              class="w-fit"
            >
              <button
                class="secondary"
                data-testid="17:00"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full flex-col justify-start gap-y-[48px] md:gap-y-[80px]"
    >
      <p
        class="font-baskervald text-[1.625rem] md:text-[3.125rem]"
      >
        Saturday, 6 JULY 2024
      </p>
      <div
        class="flex w-full"
      >
        <div
          class="ml-auto gap-x-8 gap-y-[16px] grid grid-flow-col md:grid-rows-none md:ml-0 md:grid-flow-row md:grid-cols-5 md:w-full grid-rows-5"
        >
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="08:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="09:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="10:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="11:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="12:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="13:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="14:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="15:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="16:00"
            />
          </div>
          <div
            class="w-fit"
          >
            <button
              class="secondary"
              data-testid="17:00"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
